File: Den.Dev.Grunt/Den.Dev.Grunt/Core/Modules/AcademyModule.cs
// <copyright file="AcademyModule.cs" company="Den Delimarsky">
// Developed by Den Delimarsky.
// Den Delimarsky licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
// The underlying API powering Den.Dev.Grunt is managed by Halo Studios and Microsoft. This wrapper is not endorsed by Halo Studios or Microsoft.
// </copyright>

using System.Threading.Tasks;
using Den.Dev.Grunt.Core.Foundation;
using Den.Dev.Grunt.Endpoints;
using Den.Dev.Grunt.Models;
using Den.Dev.Grunt.Models.HaloInfinite;

namespace Den.Dev.Grunt.Core.Modules
{
    /// <summary>
    /// Module for Academy-related API operations including bot customization and drills.
    /// </summary>
    public class AcademyModule : ModuleBase
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="AcademyModule"/> class.
        /// </summary>
        /// <param name="client">The client instance to use for API requests.</param>
        internal AcademyModule(ClientBase client)
            : base(client, HaloCoreEndpoints.GameCmsOrigin)
        {
        }

        /// <summary>
        /// Get bot customization information.
        /// </summary>
        /// <include file='../../APIDocsExamples/HaloInfinite/Academy_GetBotCustomization.xml' path='example'/>
        /// <param name="flightId">ID of the flight/clearance associated with the request.</param>
        /// <returns>If successful, returns an instance of BotCustomizationData that contains bot customization information. Otherwise, returns null.</returns>
        public async Task<HaloApiResultContainer<BotCustomizationData, RawResponseContainer>> GetBotCustomization(string flightId)
        {
            return await this.GetAsync<BotCustomizationData>(
                $"/hi/multiplayer/file/Academy/BotCustomizationData.json?flight={flightId}",
                useClearance: true);
        }

        /// <summary>
        /// Gets the client manifest for the Academy.
        /// </summary>
        /// <include file='../../APIDocsExamples/HaloInfinite/Academy_GetContent.xml' path='example'/>
        /// <returns>If successful, returns an instance of AcademyClientManifest that contains the definition of drills available in the Academy. Otherwise, returns null.</returns>
        public async Task<HaloApiResultContainer<AcademyClientManifest, RawResponseContainer>> GetContent()
        {
            return await this.GetAsync<AcademyClientManifest>(
                "/hi/multiplayer/file/Academy/AcademyClientManifest.json",
                useClearance: true);
        }

        /// <summary>
        /// Gets the client manifest for the Academy. From the endpoint name we can infer that this is test data.
        /// </summary>
        /// <include file='../../APIDocsExamples/HaloInfinite/Academy_GetContentTest.xml' path='example'/>
        /// <param name="clearanceId">ID of the flight/clearance associated with the request.</param>
        /// <returns>If successful, returns an instance of TestAcademyClientManifest that contains the definition of drills available in the Academy. Otherwise, returns null.</returns>
        public async Task<HaloApiResultContainer<TestAcademyClientManifest, RawResponseContainer>> GetContentTest(string clearanceId)
        {
            return await this.GetAsync<TestAcademyClientManifest>(
                $"/hi/multiplayer/file/Academy/AcademyClientManifest_Test.json?flight={clearanceId}");
        }

        /// <summary>
        /// Gets definitions for stars awarded in the Academy. This call breaks if a user agent is specified.
        /// </summary>
        /// <include file='../../APIDocsExamples/HaloInfinite/Academy_GetStarDefinitions.xml' path='example'/>
        /// <returns>If successful, returns an instance of AcademyStarDefinitions that contains definitions for stars awarded in the Academy. Otherwise, returns null.</returns>
        public async Task<HaloApiResultContainer<AcademyStarDefinitions, RawResponseContainer>> GetStarDefinitions()
        {
            return await this.GetAsync<AcademyStarDefinitions>(
                "/hi/multiplayer/file/Academy/AcademyStarGUIDDefinitions.json",
                useClearance: true);
        }
    }
}

File: src/node/src/modules/halo-infinite/academy.module.ts
import { ModuleBase } from '../base/module-base';
import type { ClientBase } from '../../clients/base/client-base';
import type { HaloApiResult } from '../../models/common/api-result';
import { HALO_CORE_ENDPOINTS } from '../../endpoints/halo-core-endpoints';
import type {
  AcademyClientManifest,
  AcademyStarDefinitions,
  BotCustomizationData,
  TestAcademyClientManifest,
} from '../../models/halo-infinite/misc';

/**
 * Academy module for bot customization and drill-related APIs.
 *
 * Provides access to:
 * - Academy content manifest
 * - Bot customization options
 * - Star/scoring definitions for drills
 *
 * @example
 * ```typescript
 * // Get academy content
 * const content = await client.academy.getContent();
 *
 * // Get bot customization options
 * const bots = await client.academy.getBotCustomization('flight-id');
 * ```
 */
export class AcademyModule extends ModuleBase {
  constructor(client: ClientBase) {
    super(client, HALO_CORE_ENDPOINTS.GAME_CMS_ORIGIN);
  }

  /**
   * Get bot customization data.
   *
   * @param flightId - Flight/clearance ID
   * @returns Bot customization options
   */
  getBotCustomization(flightId?: string): Promise<HaloApiResult<BotCustomizationData>> {
    const flightParam = flightId ? `?flight=${flightId}` : '';
    return this.get<BotCustomizationData>(`/hi/academy/botcustomization${flightParam}`, {
      useClearance: !!flightId,
    });
  }

  /**
   * Get academy content manifest.
   *
   * @returns Academy client manifest
   */
  getContent(): Promise<HaloApiResult<AcademyClientManifest>> {
    return this.get<AcademyClientManifest>('/hi/academy/content');
  }

  /**
   * Get test academy content (for flighted builds).
   *
   * @param clearanceId - Clearance identifier
   * @returns Test academy manifest
   */
  getContentTest(clearanceId: string): Promise<HaloApiResult<TestAcademyClientManifest>> {
    this.assertNotEmpty(clearanceId, 'clearanceId');
    return this.get<TestAcademyClientManifest>(`/hi/academy/content/test?clearanceId=${clearanceId}`, {
      useClearance: true,
    });
  }

  /**
   * Get star/scoring definitions for academy drills.
   *
   * @returns Star definitions
   */
  getStarDefinitions(): Promise<HaloApiResult<AcademyStarDefinitions>> {
    return this.get<AcademyStarDefinitions>('/hi/Progression/file/academy/stars');
  }
}
